der/src/reader/pem.rs

```diff
@@ -51,31 +170,39 @@ impl<'i> Reader<'i> for PemReader<'i> {
     }
 
     fn peek_byte(&self) -> Option<u8> {
-        // TODO(tarcieri): lookahead buffer
-        None
+        if self.is_finished() {
+            None
+        } else {
+            self.reader.borrow().peek_byte()
+        }
     }
 
     fn peek_header(&self) -> Result<Header> {
-        // TODO(tarcieri): lookahead buffer
-        Err(ErrorKind::Reader.into())
+        if self.is_finished() {
+            Err(Error::incomplete(self.offset()))
+        } else {
+            Header::decode(&mut self.clone())
+        }
     }
 
     fn position(&self) -> Length {
         self.position
     }
 
     fn read_slice(&mut self, _len: Length) -> Result<&'i [u8]> {
-        // Can't borrow from PEM because it requires decoding
+        // Can't borrow from PEM because it requires decoding or a that the Length parameter
+        // is smaller that our buffered size
```

Comment:
Are you trying to make it so it can borrow out of the buffer in certain cases?

I think that's going to have pretty strange and unreliable behavior.

The `DecodeOwned` bound means types have to take ownership of the data anyway, so I think it makes a lot more sense to just force them to read data into their own buffers.

Reply:
yeah no, that comment was idiotic. The `'i` lifetime doesn't work here. Please disregard.